File: checkoutdesignator/routers/preorders.py
from typing import Optional

from fastapi import APIRouter, Depends, Query
from sqlmodel import Session

from ..api import raise_http_error
from ..dependencies import db_session
from ..exceptions import CardPosError
from ..models import PreorderClaim, PreorderItem, PreorderOrder
from ..schemas import (
    PreorderClaimCreate,
    PreorderClaimFulfillRequest,
    PreorderClaimPaymentUpdate,
    PreorderClaimUpdate,
    PreorderItemCreate,
    PreorderItemCreateWithProduct,
    PreorderItemUpdate,
    PreorderReleaseRequest,
    PreorderSetCreate,
)
from ..services import preorders as preorder_service

router = APIRouter(prefix="/preorders", tags=["preorders"])


@router.get("/items", response_model=list[PreorderItem])
def list_items(session: Session = Depends(db_session)) -> list[PreorderItem]:
    return preorder_service.list_preorder_items(session)


@router.get("/claims", response_model=list[PreorderClaim])
def list_claims(
    is_paid: Optional[bool] = Query(None, description="Filter by payment status"),
    session: Session = Depends(db_session),
) -> list[PreorderClaim]:
    return preorder_service.list_preorder_claims(session, is_paid=is_paid)


@router.get("/orders", response_model=list[PreorderOrder])
def list_orders(session: Session = Depends(db_session)) -> list[PreorderOrder]:
    """List all preorder orders (groups of claims by customer)."""
    return preorder_service.list_preorder_orders(session)


@router.post("/items", response_model=PreorderItem, status_code=201)
def create_preorder_item(payload: PreorderItemCreate, session: Session = Depends(db_session)) -> PreorderItem:
    try:
        preorder_item = preorder_service.create_preorder_item(session, payload)
    except CardPosError as exc:
        raise_http_error(exc)
    return preorder_item


@router.post("/items/with-product", response_model=PreorderItem, status_code=201)
def create_preorder_with_product(
    payload: PreorderItemCreateWithProduct,
    session: Session = Depends(db_session),
) -> PreorderItem:
    """Create a preorder item along with a new inventory item for unreleased products."""
    try:
        preorder_item = preorder_service.create_preorder_with_product(session, payload)
    except CardPosError as exc:
        raise_http_error(exc)
    return preorder_item


@router.post("/sets", response_model=list[PreorderItem], status_code=201)
def create_preorder_set(
    payload: PreorderSetCreate,
    session: Session = Depends(db_session),
) -> list[PreorderItem]:
    """Create multiple preorder items with shared game and release date.
    
    This is useful for setting up an entire product release at once.
    """
    try:
        preorder_items = preorder_service.create_preorder_set(session, payload)
    except CardPosError as exc:
        raise_http_error(exc)
    return preorder_items


@router.patch("/items/{item_id}", response_model=PreorderItem)
def update_preorder_item(
    item_id: int,
    payload: PreorderItemUpdate,
    session: Session = Depends(db_session),
) -> PreorderItem:
    try:
        preorder_item = preorder_service.update_preorder_item(session, item_id, payload)
    except CardPosError as exc:
        raise_http_error(exc)
    return preorder_item


@router.post("/items/{item_id}/release", response_model=PreorderItem)
def release_preorder_to_inventory(
    item_id: int,
    payload: PreorderReleaseRequest,
    session: Session = Depends(db_session),
) -> PreorderItem:
    """Release preorder inventory to main inventory after release date."""
    try:
        preorder_item = preorder_service.release_preorder_to_inventory(session, item_id, payload)
    except CardPosError as exc:
        raise_http_error(exc)
    return preorder_item


@router.post("/claims", response_model=PreorderClaim, status_code=201)
def create_preorder_claim(payload: PreorderClaimCreate, session: Session = Depends(db_session)) -> PreorderClaim:
    try:
        claim = preorder_service.create_preorder_claim(session, payload)
    except CardPosError as exc:
        raise_http_error(exc)
    return claim


@router.patch("/claims/{claim_id}", response_model=PreorderClaim)
def update_preorder_claim(
    claim_id: int,
    payload: PreorderClaimUpdate,
    session: Session = Depends(db_session),
) -> PreorderClaim:
    try:
        claim = preorder_service.update_preorder_claim(session, claim_id, payload)
    except CardPosError as exc:
        raise_http_error(exc)
    return claim


@router.post("/claims/{claim_id}/payment", response_model=PreorderClaim)
def record_preorder_payment(
    claim_id: int,
    payload: PreorderClaimPaymentUpdate,
    session: Session = Depends(db_session),
) -> PreorderClaim:
    """Record payment for a preorder claim. Payment data is stored permanently."""
    try:
        claim = preorder_service.record_preorder_payment(session, claim_id, payload)
    except CardPosError as exc:
        raise_http_error(exc)
    return claim


@router.post("/claims/{claim_id}/cancel", response_model=PreorderClaim)
def cancel_claim(claim_id: int, session: Session = Depends(db_session)) -> PreorderClaim:
    try:
        claim = preorder_service.cancel_preorder_claim(session, claim_id)
    except CardPosError as exc:
        raise_http_error(exc)
    return claim


@router.post("/claims/{claim_id}/fulfill", response_model=PreorderClaim)
def fulfill_claim(
    claim_id: int,
    payload: PreorderClaimFulfillRequest,
    session: Session = Depends(db_session),
) -> PreorderClaim:
    try:
        claim = preorder_service.fulfill_preorder_claim(session, claim_id, payload)
    except CardPosError as exc:
        raise_http_error(exc)
    return claim


@router.post("/release-expired", response_model=dict)
def release_expired_preorders(session: Session = Depends(db_session)) -> dict:
    """
    Release all preorder allocations for items past their release date.
    This moves allocated inventory to available inventory.
    Should be run periodically (e.g., daily).
    """
    try:
        released_items = preorder_service.release_expired_preorders(session)
    except CardPosError as exc:
        raise_http_error(exc)
    return {
        "released_count": len(released_items),
        "preorder_ids": [item.id for item in released_items],
    }

File: frontend/src/api/preorders.ts
import { apiClient } from "./client";

// Preorder Item interfaces
export interface PreorderItem {
  id: number;
  inventory_item_id: number;
  release_date: string | null;
  preorder_quantity: number;
  preorder_quantity_allocated: number;
  quantity_cap: number | null;  // Keep for backward compatibility, always 1
  notes: string | null;
  created_at: string;
  updated_at: string;
  inventory_item?: any; // Can expand with full InventoryItem interface if needed
}

export interface PreorderItemCreate {
  inventory_item_id: number;
  release_date?: string;
  preorder_quantity: number;
  notes?: string;
}

export interface PreorderItemCreateWithProduct {
  product_name: string;
  sku: string;
  game_title: string;
  category?: string;
  msrp_cents?: number;
  release_date?: string;
  preorder_quantity: number;
  notes?: string;
}

export interface PreorderItemUpdate {
  release_date?: string;
  preorder_quantity?: number;
  notes?: string;
  msrp_cents?: number;
  set_code?: string;
}

export interface PreorderSetProduct {
  product_name: string;
  sku: string;
  msrp_cents?: number;
  preorder_quantity: number;
}

export interface PreorderSetCreate {
  game_title: string;
  release_date?: string;
  category?: string;
  notes?: string;
  products: PreorderSetProduct[];
}

// Preorder Claim interfaces
export interface PreorderClaim {
  id: number;
  preorder_order_id: number;
  preorder_item_id: number;
  customer_id: number;
  quantity_requested: number;
  quantity_allocated: number;
  status: string;
  is_paid: boolean;
  payment_amount_cents: number | null;
  payment_method: string | null;
  payment_date: string | null;
  payment_notes: string | null;
  created_at: string;
  updated_at: string;
  preorder_item?: PreorderItem;
  customer?: any; // Can expand with full Customer interface if needed
}

// Preorder Order interface (groups claims by customer)
export interface PreorderOrder {
  id: number;
  customer_id: number;
  status: string;
  notes: string | null;
  created_at: string;
  updated_at: string;
  customer?: any;
  claims: PreorderClaim[];
}

export interface PreorderClaimCreate {
  preorder_item_id: number;
  customer_id: number;
  quantity_requested: number;
}

export interface PreorderClaimUpdate {
  quantity_requested?: number;
  quantity_allocated?: number;
  status?: string;
}

export interface PreorderClaimPaymentUpdate {
  is_paid: boolean;
  payment_amount_cents: number;
  payment_method: string;
  payment_notes?: string;
}

export interface PreorderClaimFulfillRequest {
  mark_picked_up?: boolean;
  note?: string;
}

export interface PreorderReleaseRequest {
  note?: string;
}

// Preorder Items API
export const fetchPreorderItems = async (): Promise<PreorderItem[]> => {
  const response = await apiClient.get("/preorders/items");
  return response.data;
};

export const createPreorderItem = async (item: PreorderItemCreate): Promise<PreorderItem> => {
  const response = await apiClient.post("/preorders/items", item);
  return response.data;
};

export const createPreorderWithProduct = async (item: PreorderItemCreateWithProduct): Promise<PreorderItem> => {
  const response = await apiClient.post("/preorders/items/with-product", item);
  return response.data;
};

export const createPreorderSet = async (set: PreorderSetCreate): Promise<PreorderItem[]> => {
  const response = await apiClient.post("/preorders/sets", set);
  return response.data;
};

export const updatePreorderItem = async (
  itemId: number,
  update: PreorderItemUpdate
): Promise<PreorderItem> => {
  const response = await apiClient.patch(`/preorders/items/${itemId}`, update);
  return response.data;
};

export const releasePreorderToInventory = async (
  itemId: number,
  request: PreorderReleaseRequest
): Promise<PreorderItem> => {
  const response = await apiClient.post(`/preorders/items/${itemId}/release`, request);
  return response.data;
};

// Preorder Claims API
export const fetchPreorderClaims = async (isPaid?: boolean): Promise<PreorderClaim[]> => {
  const params = isPaid !== undefined ? { is_paid: isPaid } : {};
  const response = await apiClient.get("/preorders/claims", { params });
  return response.data;
};

// Preorder Orders API (grouped claims by customer)
export const fetchPreorderOrders = async (): Promise<PreorderOrder[]> => {
  const response = await apiClient.get("/preorders/orders");
  return response.data;
};

export const createPreorderClaim = async (claim: PreorderClaimCreate): Promise<PreorderClaim> => {
  const response = await apiClient.post("/preorders/claims", claim);
  return response.data;
};

export const updatePreorderClaim = async (
  claimId: number,
  update: PreorderClaimUpdate
): Promise<PreorderClaim> => {
  const response = await apiClient.patch(`/preorders/claims/${claimId}`, update);
  return response.data;
};

export const recordPreorderPayment = async (
  claimId: number,
  payment: PreorderClaimPaymentUpdate
): Promise<PreorderClaim> => {
  const response = await apiClient.post(`/preorders/claims/${claimId}/payment`, payment);
  return response.data;
};

export const cancelPreorderClaim = async (claimId: number): Promise<PreorderClaim> => {
  const response = await apiClient.post(`/preorders/claims/${claimId}/cancel`);
  return response.data;
};

export const fulfillPreorderClaim = async (
  claimId: number,
  request: PreorderClaimFulfillRequest
): Promise<PreorderClaim> => {
  const response = await apiClient.post(`/preorders/claims/${claimId}/fulfill`, request);
  return response.data;
};

export const unfulfillPreorderClaim = async (claimId: number): Promise<PreorderClaim> => {
  const response = await apiClient.post(`/preorders/claims/${claimId}/unfulfill`);
  return response.data;
};
